### PR TITLE
MINOR: removed code duplicates from several files (KafkaStreams)

### PR DESCRIPTION
**Pull request by wlsc**
This PR offers following to KafkaStreams project:
* removed code duplication from **TopologyBuilder**'s function **makeNodeGroups** (extract function)
* removed code duplication from **GlobalProcessorContextImpl** and from **ProcessorContextImpl** to parent class **AbstractProcessorContext** (move method to parent)

### CODE REVIEW

**Comment by wlsc** [comment]
Any updates on this PR?

**Comment by guozhangwang** [comment]
@wlsc sorry for being late on reviews, could you rebase your code on latest trunk?

**Comment by wlsc** [comment]
@guozhangwang hi. thank you for the review, but my copy/branch of Kafka was already deleted. 
If you don't mind, I could create another PR and link it here. 
Would be that ok for you?

**Comment by wlsc** [comment]
@guozhangwang I've created another PR for this. Please check the https://github.com/apache/kafka/pull/4357 PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java [hunk 1]
```diff
@@ -1115,43 +1115,41 @@ private void connectStateStoreNameToSourceTopicsOrPattern(final String stateStor
...
-                }
-                nodeGroup.add(nodeName);
+                nodeGroupId = getNodeGroupId(nodeGroups, rootToNodeGroup, nodeGroupId, nodeName);
             }
         }
 
         return nodeGroups;
     }
 
+    private int getNodeGroupId(Map<Integer, Set<String>> nodeGroups, Map<String, Set<String>> rootToNodeGroup, int nodeGroupId, String nodeName) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:1141]: nit: add `final` to each parameter and use single parameter per line formatting

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java [hunk 2]
```diff
@@ -100,6 +101,20 @@ public void register(final StateStore store, final boolean loggingEnabled, final
         stateManager.register(store, loggingEnabled, stateRestoreCallback);
     }
 
+    @SuppressWarnings("unchecked")
+    @Override
+    public <K, V> void forward(K key, V value) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:106]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java [hunk 3]
```diff
@@ -100,6 +101,20 @@ public void register(final StateStore store, final boolean loggingEnabled, final
...
+        final ProcessorNode previousNode = currentNode();
+        try {
+            for (ProcessorNode child : (List<ProcessorNode>) currentNode().children()) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:109]: nit: add `final`

**Review by mjsax — COMMENTED** [review]
Just some nits and further improvements.

**Comment by wlsc** [comment]
@mjsax sorry, I had to close this actually. I will do your PR changes also in the second PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java (hunk 3 above)
Comment (wlsc) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:109]: ok

File: streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java (hunk 1 above)
Comment (wlsc) [reply streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:1141]: ok

**Pull request closed without merging**